### ISSUE TITLE
[discussion] Remove the need of mandatory super() module call

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#61686]
I think I already made this proposal somewhere in the comments some time ago, but I could not find it.

It seems that the default constructor only adding fields (like hooks), and these fields could be made into lazy properties, so that underlying fields are initialised when the property is first called.

If this is indeed the only thing to be done, it could make the user code more flexible.

**Comment by jbschlosser** [comment]
Hey @vadimkantorov, thanks for the suggestion! While I agree it'd ideally be nice to remove the mandatory call, it seems like a relatively low UX benefit compared to the effort / risk required in taking this on (if it's feasible to do).

**Comment by vadimkantorov** [comment]
I agree it may be brittle, but it would enable scenarios like https://github.com/pytorch/pytorch/issues/58997#issuecomment-851025969 - integrating with existing almost-module code.

Still worth technical discussion I think (even if decided against at the end) - examining carefully what is done in existing Module's init and what parts can be safely moved elsewhere.

**Comment by albanD** [comment]
Note that I don't think we will be able to make this call lazy :/
https://github.com/pytorch/pytorch/blob/094abf5fd0d2b1efa77bfe88b036c677b790d800/torch/nn/modules/module.py#L254

**Comment by vadimkantorov** [comment]
what's function of this call? if it's just logging, maybe it's not a blocker. the rest of initialization could probably be put into getattr

**Comment by vadimkantorov** [comment]
if this is supported, then modules can also be dataclasses and have `__init__` generated automatically - and for simple parameter fields, this would make terser code

**Comment by vadimkantorov** [comment]
(also if a `__call__` is supported more natively (even if disabling the hooks; or maybe even supported now), it would make PyTorch more close to Jax)

**Comment by vadimkantorov** [comment]
Also, for discussion: allow "duck-typed" modules without any base class and using just `__call__` without forward (understood that it would break hooks). Motivation: helper modules like in https://github.com/pytorch/vision/pull/4389 + this is used in Jax (?).

I think it would break many things in module APIs, but could it be worked around? are there any contexts where this would make sense?

**Comment by albanD** [comment]
Well, the benefit of have them Module subclass is that we can check what they are and have some shared method on them (to get params and submodules for example).

> are there any contexts where this would make sense?

I'm not sure to see any use case where it would be better not to have them subclasses. Do you have one?

**Comment by vadimkantorov** [comment]
I think it's mostly for porting / reusing existing code without having to do sweeping fixes. This will become more important with more unification of APIs between PyTorch / NumPy etc.

I agree that it's useful to have the shared methods, but I guess accessing many of them can be done in duck-typing way.

I guess practicality would be descending: alleviating super().__init__() call; supporting __call__; alleviating need for inheritance

**Comment by vadimkantorov** [comment]
Also, one argument, this lets you create instances but can't do forward:

```python
class IdentityBlock(nn.Identity):
  def __init__(self, **kwargs):
    pass

IdentityBlock()(torch.rand(3))
```

**Comment by yassersouri** [comment]
Another argument for removing the need to explicitly call `super().__init__()` is that currently Modules cannot be a dataclass.

For example the following is not possible:
```python
@dataclass()
class Net(nn.Module):
    other_layer: nn.Module
    input_feats: int = 10
    output_feats: int = 20

    def __post_init__(self):
        self.layer = nn.Linear(self.input_feats, self.output_feats)

    def forward(self, x):
        return self.layer(self.other_layer(x))
```

One cannot simply call `super().__init__()` in the `__post_init__` function.
More discussion [here](https://discuss.pytorch.org/t/how-to-use-dataclass-with-pytorch/53444/4)

**Comment by enhuiz** [comment]
> Another argument for removing the need to explicitly call `super().__init__()` is that currently Modules cannot be a dataclass.
> 
> For example the following is not possible:
> 
> ```python
> @dataclass()
> class Net(nn.Module):
>     other_layer: nn.Module
>     input_feats: int = 10
>     output_feats: int = 20
> 
>     def __post_init__(self):
>         self.layer = nn.Linear(self.input_feats, self.output_feats)
> 
>     def forward(self, x):
>         return self.layer(self.other_layer(x))
> ```
> 
> One cannot simply call `super().__init__()` in the `__post_init__` function. More discussion [here](https://discuss.pytorch.org/t/how-to-use-dataclass-with-pytorch/53444/4)

One workaround can be (perhaps not very elegant):

```python
import torch.nn as nn
from dataclasses import dataclass
from typing import Callable


@dataclass()
class Net(nn.Module):
    other_layer_creater: Callable[[], nn.Module]
    input_feats: int = 10
    output_feats: int = 20

    def __post_init__(self):
        super().__init__()
        self.layer = nn.Linear(self.input_feats, self.output_feats)
        self.other_layer = self.other_layer_creater()

    def forward(self, x):
        return self.layer(self.other_layer(x))


Net(other_layer_creater=lambda: nn.Linear(10, 10))
```

**Comment by rsokl** [comment]
This is a potential solution to these issues:

```python
import torch as tr
import torch.nn as nn
from dataclasses import dataclass

@dataclass
class DataclassModule(nn.Module):
    def __new__(cls, *args, **k):
        inst = super().__new__(cls)
        nn.Module.__init__(inst)
        return inst

@dataclass(unsafe_hash=True)
class Net(DataclassModule):
    other_layer: nn.Module
    input_feats: int = 10
    output_feats: int = 20

    def __post_init__(self):
        self.layer = nn.Linear(self.input_feats, self.output_feats)

    def forward(self, x):
        return self.layer(self.other_layer(x))

net = Net(other_layer=nn.Linear(10, 10))
assert net(tr.tensor([1.]*10)).shape == (20,)
assert len(list(net.parameters())) == 4
```

**Comment by vadimkantorov** [comment]
Or maybe could this `__new__` be added directly to `nn.Module`? like so it could do attribute initialization there, and not require it be done in derived class init?

Then the question is whether this scheme would work correctly with derived class being decorated as `@dataclass`
